### PR TITLE
chore(cli): advice customers on netmask dependency

### PR DESCRIPTION
**Pull request by nija-at**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk/bin/cdk.ts [hunk 1]
```diff
@@ -262,6 +264,7 @@ async function initCommandLine() {
         // In code it's optimistically called "default" bootstrapping but that is in
         // anticipation of flipping the switch, in user messaging we still call it
         // "new" bootstrapping.
+        printAdvisories();
```
Comment (rix0rrr) [inline packages/aws-cdk/bin/cdk.ts:267]: I wouldn't do it here.

File: packages/aws-cdk/lib/advisory.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+interface Advisory {
+  message(): string | void;
+}
+
+class NetmaskAdvisory implements Advisory {
+  public message(): string | void {
+    const pkgLockJsonPath = path.join(process.cwd(), 'package-lock.json');
+    const pkgJsonPath = path.join(process.cwd(), 'package.json');
+
+    if (fs.existsSync(pkgLockJsonPath)) {
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/advisory.ts:26]: Doesn't it make more sense to check for a `yarn.lock` instead to DO do the work, rather than the presence of a `package-lock.json` to NOT do the work?

**Comment by rix0rrr** [comment]
Good enough for now I suppose.

But this doesn't scale, and is a poor user experience when enabling and disabling advisories. I'd still vote for some SIMPLE, but existing, dynamic configuration fetching.

We've been talking about this for a while, and we're constantly punting it because it's "too hard/needs more design".

Happy to throw up a straw man design.

File: packages/aws-cdk/lib/advisory.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    const pkgLockJsonPath = path.join(process.cwd(), 'package-lock.json');
+    const pkgJsonPath = path.join(process.cwd(), 'package.json');
+
+    if (!fs.existsSync(pkgJsonPath) || fs.existsSync(pkgLockJsonPath)) {
+      return;
+    }
+    // eslint-disable-next-line @typescript-eslint/no-require-imports
+    const json = require(pkgJsonPath);
+    const netmaskVer = json?.resolutions?.netmask;
+    if (typeof(netmaskVer) === 'string' && netmaskVer.startsWith('2')) {
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/advisory.ts:33]: Your proposed advisory wording is also making me realize we should check NPM users for a CLI version `< (whatever number)`.

Comment (rix0rrr) [inline packages/aws-cdk/lib/advisory.ts:33]: We should also only do this if `aws-cdk` is in either `dependencies` (incorrect, but possible) or `devDependencies`.

**Comment by rix0rrr** [comment]
Also, CLIs <1.96.0 should print this advisory.

File: packages/aws-cdk/lib/advisory.ts (hunk 3 above)
Comment (nija-at) [reply packages/aws-cdk/lib/advisory.ts:33]: Given that we're releasing this messaging in the next version, they would only see this message once they've upgraded. Does this make that message redundant?

**Comment by iliapolo** [comment]
@nija-at This PR doesn't address scenarios where the CLI was installed globally via yarn. Granted, this might be an edge case, but it might be that yarn users are also accustomed to doing `yarn global add` instead of `npm install -g`. 

Also, as already mentioned, this doesn't address non JS users. Do you we have a plan for those users? 

I wanted to suggest an alternative, that **might** address all use-cases. Could we use some `require` trickery to detect the actual version of `netmask` that our CLI uses, and simply display the message if the version is not > `2`? I think it should be possible since we know the exact dependency chain it comes up, allowing us to traverse `node_modules` until we find it. 

I know there might be some gotchas with respect to where exactly the dependency lands after npm/yarn resolutions. But my intuition says there is a way to deterministically detect it. 

Basically a something like:

```ts
try {
  const netmaskVersion = require(path.join(require.resolve('netmask'), '..', '..', 'package.json'))
} catch {
  // recurse until we exhaust the dependency chain.
}
```

Thoughts?

**Comment by nija-at** [comment]
> Granted, this might be an edge case, but it might be that yarn users are also accustomed to doing yarn global add instead of npm install -g

> this doesn't address non JS users. Do we have a plan for those users?

In both these cases, users have to upgrade to the latest CDK version, once we release it. And they should be good to go.

RE: `require` trickery - feels a bit too hacky for my taste, but maybe it'll work. 
Do you want to try and submit a commit or a PR on top of my PR to see how this will look?

**Comment by nija-at** [comment]
>> Granted, this might be an edge case, but it might be that yarn users are also accustomed to doing yarn global add instead of npm install -g
>
>> this doesn't address non JS users. Do we have a plan for those users?
>
> In both these cases, users have to upgrade to the latest CDK version, once we release it. And they should be good to go.

Actually, I take it back in the case of yarn global install. This needs to be verified.

**Comment by iliapolo** [comment]
> Also, as already mentioned, this doesn't address non JS users. Do we have a plan for those users?

Sorry, this actually does address those users, assuming they use `npm` and not `yarn`, which is probably a fair assumption. 

Nevertheless, I whipped up this code, which doesn't assume anything, simply detects the actual version of netmask being used:

`detect-netmask.js` is placed in `node_modules/aws-cdk/lib` (i.e where our CLI runs from):

```js
const path = require('path');
const cwd = process.cwd()
try {
  for (const dep of ['proxy-agent', 'pac-proxy-agent', 'pac-resolver']) {
    const depRoot = path.dirname(require.resolve(dep));
    process.chdir(depRoot);
  }
  const netmaskLib = path.dirname(require.resolve('netmask'));
  const netmaskRoot = path.join(path.dirname(netmaskLib));
  const netmaskPackageJson = path.join(netmaskRoot, 'package.json');
  console.log(require(netmaskPackageJson).version)
} finally {
  process.chdir(cwd)
}
```

This produces:

```console
❯ node detect-netmask.js                                                                                                                                                            [14:20:46]
1.0.6
```

It basically simulates the require statements that will happen during the proxy initialization.

@nija-at @rix0rrr thoughts?

**Comment by nija-at** [comment]
superceded by #13914
